backends/opt/ceed-opt-operator.c

```diff
@@ -59,12 +59,27 @@ static int CeedOperatorSetupFields_Opt(CeedQFunction qf, CeedOperator op, bool i
         CeedCallBackend(
             CeedElemRestrictionCreateBlockedStrided(ceed, num_elem, elem_size, blk_size, num_comp, l_size, strides, &blk_restr[i + start_e]));
       } else {
-        const CeedInt *offsets = NULL;
+        const CeedInt *offsets      = NULL;
+        const bool    *orients      = NULL;
+        const CeedInt *curl_orients = NULL;
         CeedCallBackend(CeedElemRestrictionGetOffsets(r, CEED_MEM_HOST, &offsets));
+        CeedCallBackend(CeedElemRestrictionGetOrientations(r, CEED_MEM_HOST, &orients));
+        CeedCallBackend(CeedElemRestrictionGetCurlOrientations(r, CEED_MEM_HOST, &curl_orients));
         CeedCallBackend(CeedElemRestrictionGetCompStride(r, &comp_stride));
-        CeedCallBackend(CeedElemRestrictionCreateBlocked(ceed, num_elem, elem_size, blk_size, num_comp, comp_stride, l_size, CEED_MEM_HOST,
-                                                         CEED_COPY_VALUES, offsets, &blk_restr[i + start_e]));
+        if (!orients && !curl_orients) {
```

Comment:
This makes me wonder if we need an enum for restriction types to make sure backends don't miss a case?

Reply:
That was another approach, similar to the `CeedFESpace` variable for a basis. What do you think? I'm not sure how much more I'd expect things to expand in terms of variants of restriction types.

Reply:
I could go either way. I like the enums because we have different impls of the same logic and I like the compiler yelling at me when I forget a case. I would picture this being used internally only.

Reply:
Check out 3b8f45e for a suggestion of how this might look with enums. I'm not sure which is better, open to input.

Reply:
I think it makes more sense to me that way, personally? I'd be curious to see what others think.

Reply:
Note - small consistency update for creating Blocked restrictions here: https://github.com/CEED/libCEED/pull/1205

Reply:
I agree, it it nice this way with the enum. Consistency update fixed after rebase in [27f40bc](https://github.com/CEED/libCEED/pull/1168/commits/27f40bc230702d5d2f53964f310f001318da0e54).

Reply:
I don't think the unsigned apply should add new type of restriction to the enum, and instead think it as the following (this is what I've been working with locally where I've merged this branch with #1196):
- `CEED_RESTRICTION_DEFAULT` and `CEED_RESTRICTION_STRIDED`: Have apply, unsigned apply methods which are identical
- `CEED_RESTRICTION_ORIENTED`: Has a special unsigned apply which ignores the orientations and performs the same apply as `CEED_RESTRICTION_DEFAULT`
- `CEED_RESTRICTION_CURL_ORIENTED`: Has a special unsigned apply which applies |E| or |E|^T, this results in a convergent diagonal and is similar to what MFEM does for non-conforming meshes where the P (parallel prolongation) matrix has potentially multiple entries per row (https://github.com/mfem/mfem/blob/master/fem/pbilinearform.cpp#L309).

Reply:
Maybe its the word "Default" that feels strange to me. If the restriction is set up as oriented, the default apply method should be "CEED_RESTRICTION_ORIENTED"

Reply:
Is this resolved?

Reply:
@jeremylt if you are happy with the current implementation, then it's resolved. Just let me know if you have any questions about it or see things I should change. The enum values are still:

```
/// Type of element restriction;
/// @ingroup CeedElemRestriction
typedef enum {
  /// Default element restriction with offsets
  CEED_RESTRICTION_DEFAULT = 1,
  /// Oriented element restriction
  CEED_RESTRICTION_ORIENTED = 2,
  /// Curl-oriented element restriction
  CEED_RESTRICTION_CURL_ORIENTED = 3,
  /// Strided element restriction
  CEED_RESTRICTION_STRIDED = 4,
} CeedRestrictionType;
```

but hopefully the reorganization of the apply methods in `ceed-ref-restriction.c` has helped make the organization and naming more clear.

Reply:
I don't love `DEFAULT` as "not strided, oriented, or curl oriented", but I don't have a better name. It sounds funny in my mind because different `CeedElemRestriction*Create` have different "default" application techniques. Maybe `CEED_RESTRICTION_UNSIGNED`? I dunno, maybe its just a mild :bike: :house: on how the words sound in my mind against the API

I think the organization is clear though.

Reply:
I guess the best argument for `DEFAULT` is the public API. We have `CeedElemRestrictionCreate`, `...CreateOriented`, `...CreateCurlOriented`, and `...CreateStrided`, so the "default" case really is defined there by "not strided, oriented, or curl oriented" as you say. But I'm in favor of using something more descriptive. `"OFFSETTED"` is a pretty ugly past-participle, but how about `"STANDARD"` to avoid the use of "default"?

Reply:
`STANDARD` does work in my brain better. Thanks for the flexibility!

Reply:
When you call one of them `CurlOriented` why not just call the others by the FE Space that we are using? maybe`CreatedH1`, `CreatedHdiv` and `CreatedHcurl` ?

Reply:
Couldn't we just use `_OFFSET` as the type? Isn't it its own past participle? :)  
I think the argument for "standard" from function names makes a lot of sense, though.

Reply:
Nice, done.

> When you call one of them CurlOriented why not just call the others by the FE Space that we are using? maybeCreatedH1, CreatedHdiv and CreatedHcurl ?

I think it's a good idea to decouple the element restriction from the FE space and basis as much as possible. For example, we use `CreateStrided`, not `CreateL2`. Also this would involve some breaking changes to the public API that would be good to avoid.

> Couldn't we just use `_OFFSET` as the type? Isn't it its own past participle? :) I think the argument for "standard" from function names makes a lot of sense, though.

This would work too. To avoid getting too far into the bike shed regime I just went with `STANDARD` in 
82a976c. If there are no objections I'll consider this resolved.

Reply:
My only question is whether `STRIDED` is semantically an alternative to `ORIENTED`, versus orthogonal to? I probably need to read more carefully, but that part read strangely to me.

Reply:
Interesting and good point that I guess we could have an oriented, strided restriction if I understand what you are saying. But the way things are implemented they are orthogonal and with increasing levels of generality between the different types. By that I mean that a `STANDARD` restriction captures all the functionality of the (more optimized) `STRIDED` one, adding the ability to overlap points between elements. Likewise `ORIENTED` augments `STANDARD` by adding orientations (if these are all `false` we have `STRIDED` behavior), and finally `CURL_ORIENTED` augments `ORIENTED` by allowing more general element-wise transformations. So even if `STRIDED` and `ORIENTED` are not really orthogonal but could be composed, the user should just use `CeedElemRestrictionCreateOriented` with a non-overlapping `offsets` array. I hope I understood your comment correctly, let me know if this makes sense or not.

Reply:
Yeah, that was kind of my question. An enum is fine for a hierarchy of generality, but not for what may eventually become the cross product of orthogonal knobs. If this is a hierarchy, then there will be no need to later add other entries to complete a cross product.

> But the way things are implemented they are orthogonal ...

An enum (mutually exclusive options) doesn't sound like orthogonality, though I'm not sure you intended it that way.

Reply:
At least how we currently use strided, its only for when the evec the same length as the lvec and is strictly a permutation of the lvec (And I think our only current usage is for QData). (maybe we should have called it a permutation - names are hard).

Is there a use case where we will want to permute and orient? It feels like in that case we'd be building the orientation array anyways so using the current implementation wouldn't be a big difference in performance with making a special strided + oriented implementation.

Reply:
Thanks, both, for this discussion.. The present design sounds good to me.